src/uproot/source/fsspec.py

```diff
@@ -0,0 +1,126 @@
+        self._fh = self._file.__enter__()
+        return self
+
+    def __exit__(self, exception_type, exception_value, traceback):
+        self._fh = None
+        self._file.__exit__(exception_type, exception_value, traceback)
+
+    def chunk(self, start, stop):
+        """
+        Args:
+            start (int): Seek position of the first byte to include.
+            stop (int): Seek position of the first byte to exclude
+                (one greater than the last byte to include).
+
+        Request a byte range of data from the file as a
+        :doc:`uproot.source.chunk.Chunk`.
+        """
+        self._num_requests += 1
+        self._num_requested_chunks += 1
+        self._num_requested_bytes += stop - start
+        if self._fh:
+            self._fh.seek(start)
+            data = self._fh.read(stop - start)
+        else:
+            data = self._fs.cat_file(self._file_path, start, stop)
+        future = uproot.source.futures.TrivialFuture(data)
+        return uproot.source.chunk.Chunk(self, start, stop, future)
+
+    def chunks(self, ranges, notifications):
+        """
+        Args:
+            ranges (list of (int, int) 2-tuples): Intervals to fetch
+                as (start, stop) pairs in a single request, if possible.
+            notifications (``queue.Queue``): Indicator of completed
+                chunks. After each gets filled, it is ``put`` on the
+                queue; a listener should ``get`` from this queue
+                ``len(ranges)`` times.
+
+        Request a set of byte ranges from the file.
+
+        This method has two outputs:
+
+        * The method returns a list of unfilled
+          :doc:`uproot.source.chunk.Chunk` objects, which get filled
+          in a background thread. If you try to read data from an
+          unfilled chunk, it will wait until it is filled.
+        * The method also puts the same :doc:`uproot.source.chunk.Chunk`
+          objects onto the ``notifications`` queue as soon as they are
+          filled.
+
+        Reading data from chunks on the queue can be more efficient than
+        reading them from the returned list. The total reading time is the
+        same, but work on the filled chunks can be better parallelized if
+        it is triggered by already-filled chunks, rather than waiting for
+        chunks to be filled.
+        """
+        self._num_requests += 1
+        self._num_requested_chunks += len(ranges)
+        self._num_requested_bytes += sum(stop - start for start, stop in ranges)
+        data = self._fs.cat_ranges(
+            [self._file_path] * len(ranges),
+            [start for start, _ in ranges],
+            [stop for _, stop in ranges],
+        )
+        chunks = []
+        for item, (start, stop) in zip(data, ranges):
+            future = uproot.source.futures.TrivialFuture(item)
+            chunk = uproot.source.chunk.Chunk(self, start, stop, future)
+            uproot.source.chunk.notifier(chunk, notifications)()
```

Comment:
It's a blocking call in this first implementatation, but will be made asynchronous later.